core/src/main/java/com/themecleanflex/models/ReferenceModel.java

```diff
@@ -605,18 +610,66 @@ private String generateReferenceJson() {
       return null;
     }
 
-    // find a node with the given key/value pair in our json output
+	private boolean createsReferenceLoop(final Resource resource, final Set<String> forbiddenPaths) {
+    	if (forbiddenPaths.contains(resource.getPath())) {
+    		return true;
+	    }
+
+		forbiddenPaths.addAll(forbiddenReferences(resource.getPath()));
+		if (Optional.of(resource)
+					.filter(r -> r.isResourceType("themecleanflex/components/reference"))
+					.map(Resource::getValueMap)
+					.map(props -> props.get("reference", String.class))
+					.map(resource.getResourceResolver()::getResource)
+					.map(r -> createsReferenceLoop(r, forbiddenPaths))
+					.orElse(false)) {
+			return true;
+		}
+
+    	for (final Resource child : resource.getChildren()) {
+    		if (createsReferenceLoop(child, forbiddenPaths)) {
+    			return true;
+		    }
+	    }
+
+	    return false;
+	}
+
+	private static Set<String> forbiddenReferences(final String path) {
+		final Set<String> result = new HashSet<>();
+		result.add(path);
+		if (!isJcrContentDescendant(path)) {
+			if (PerUtil.isJcrContent(path)) {
+				result.add(PerUtil.getParent(path));
+			}
+
+			return result;
+		}
+
+		String parent = path;
+		do {
+			parent = PerUtil.getParent(parent);
+			result.add(parent);
+		} while (!PerUtil.isJcrContent(parent));
+		result.add(PerUtil.getParent(parent));
+		return result;
+	}
+
+	private static boolean isJcrContentDescendant(final String path) {
+		return contains(path, "/jcr:content/");
+	}
+
+	// find a node with the given key/value pair in our json output
     private Map findNode(Map map, String name, String value) {
       for (Object key : map.keySet()) {
         Object val = map.get(key);
-        if(key.equals(name)) {
-          if(value.equals(val)) {
+        if(equals(key, name)) {
+          if(equals(value, val)) {
             return map;
           }
         }
-        if(key.equals("children") && val instanceof ArrayList) {
+        if("children".equals(key) && val instanceof ArrayList) {
```

Comment:
Again the same question: why use `equals()` above and `String.equals()` here?

Reply:
`"children"` is a literal, which is never `null`, the other calls had to be made `null`-safe.

Reply:
`"children"` is a literal, which is never `null`, the other calls had to be made `null`-safe.